{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python\n",
    "\n",
    "import sys\n",
    "import gzip\n",
    "import numpy as np\n",
    "import scipy as sc\n",
    "import pickle\n",
    "\n",
    "from optparse import OptionParser\n",
    "    \n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn import preprocessing\n",
    "from sklearn import linear_model\n",
    "\n",
    "from scipy.stats import rankdata\n",
    "from scipy.stats import norm\n",
    "\n",
    "\n",
    "ratio_file = \"QTL_results/QTL_perind.counts_chrname.gz\"\n",
    "\n",
    "def qqnorm(x):\n",
    "    n=len(x)\n",
    "    a=3.0/8.0 if n<=10 else 0.5\n",
    "    return(norm.ppf( (rankdata(x)-a)/(n+1.0-2.0*a) ))\n",
    "\n",
    "def stream_table(f, ss = ''):\n",
    "    fc = '#'\n",
    "    while fc[0] == \"#\":\n",
    "        fc = f.readline().strip()\n",
    "        head = fc.split(ss)\n",
    "\n",
    "    for ln in f:\n",
    "        ln = ln.strip().split(ss)\n",
    "        attr = {}\n",
    "\n",
    "        for i in range(len(head)):\n",
    "            try: attr[head[i]] = ln[i]\n",
    "            except: break\n",
    "        yield attr\n",
    "\n",
    "def get_chromosomes(ratio_file):\n",
    "    \"\"\"Get chromosomes from table. Returns set of chromosome names\"\"\"\n",
    "    try: open(ratio_file)\n",
    "    except:\n",
    "        sys.stderr.write(\"Can't find %s..exiting\\n\"%(ratio_file))\n",
    "        return\n",
    "    sys.stderr.write(\"Parsing chromosome names...\\n\")\n",
    "    chromosomes = set()\n",
    "    with gzip.open(ratio_file, 'rt') as f:\n",
    "            f.readline()\n",
    "            for line in f:\n",
    "                chromosomes.add(line.split(\":\")[0])\n",
    "    return(chromosomes)\n",
    "\n",
    "def get_blacklist_chromosomes(chromosome_blacklist_file):\n",
    "    \"\"\"\n",
    "    Get list of chromosomes to ignore from a file with one blacklisted\n",
    "    chromosome per line. Returns list. eg. ['X', 'Y', 'MT']\n",
    "    \"\"\"\n",
    "    if chromosome_blacklist_file:\n",
    "        with open(chromosome_blacklist_file, 'r') as f:\n",
    "            return(f.read().splitlines())\n",
    "    else:\n",
    "        return([\"X\", \"Y\"])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_chromosomes(ratio_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(ratio_file, chroms, blacklist_chroms, pcs=50):\n",
    "    \n",
    "    dic_pop, fout = {}, {}\n",
    "    try: open(ratio_file)\n",
    "    except: \n",
    "        sys.stderr.write(\"Can't find %s..exiting\\n\"%(ratio_file))\n",
    "        return\n",
    "\n",
    "    sys.stderr.write(\"Starting...\\n\")\n",
    "    for i in chroms:\n",
    "        fout[i] = file(ratio_file+\".phen_chr\"+i,'w')\n",
    "        fout_ave = file(ratio_file+\".ave\",'w')\n",
    "    valRows, valRowsnn, geneRows = [], [], []\n",
    "    finished = False\n",
    "    header = gzip.open(ratio_file).readline().split()[1:]\n",
    "    \n",
    "    for i in fout:\n",
    "        fout[i].write(\"\\t\".join([\"#Chr\",\"start\", \"end\", \"ID\"]+header)+'\\n')\n",
    "        print(fout[1])\n",
    "    \n",
    "    for dic in stream_table(gzip.open(ratio_file),' '):\n",
    "\n",
    "        chrom = dic['chrom'].replace(\"chr\",'')\n",
    "        chr_ = chrom.split(\":\")[0]\n",
    "        #if chr_ in blacklist_chroms: continue\n",
    "        NA_indices, valRow, aveReads = [], [], []\n",
    "        tmpvalRow = []\n",
    "\n",
    "        i = 0\n",
    "        for sample in header:\n",
    "\n",
    "            try: count = dic[sample]\n",
    "            except: print (chrom), len(dic)\n",
    "            num, denom = count.split('/')\n",
    "            if float(denom) < 1:\n",
    "                count = \"NA\"\n",
    "                tmpvalRow.append(\"NA\")\n",
    "                NA_indices.append(i)\n",
    "            else:\n",
    "                # add a 0.5 pseudocount\n",
    "                count = (float(num)+0.5)/((float(denom))+0.5)\n",
    "                tmpvalRow.append(count) \n",
    "                aveReads.append(count)\n",
    "\n",
    "\n",
    "        # If ratio is missing for over 40% of the samples, skip\n",
    "        if tmpvalRow.count(\"NA\") > len(tmpvalRow)*0.4:\n",
    "            continue\n",
    "\n",
    "        ave = np.mean(aveReads)\n",
    "\n",
    "        # Set missing values as the mean of all values\n",
    "        for c in tmpvalRow:\n",
    "            if c == \"NA\": valRow.append(ave)\n",
    "            else: valRow.append(c)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "        # If there is too little variation, skip (there is a bug in fastqtl which doesn't handle cases with no variation)\n",
    "        if np.std(valRow) < 0.005: continue\n",
    "\n",
    "        chr_, s, e, clu = chrom.split(\":\")\n",
    "        if len(valRow) > 0:                \n",
    "            fout[chr_].write(\"\\t\".join([chr_,s,e,chrom]+[str(x) for x in valRow])+'\\n')\n",
    "            fout_ave.write(\" \".join([\"%s\"%chrom]+[str(min(aveReads)), str(max(aveReads)), str(np.mean(aveReads))])+'\\n')\n",
    "\n",
    "\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "            # scale normalize\n",
    "            valRowsnn.append(valRow)                \n",
    "            valRow = preprocessing.scale(valRow)\n",
    "\n",
    "            valRows.append(valRow)\n",
    "            geneRows.append(\"\\t\".join([chr_,s,e,chrom]))\n",
    "            if len(geneRows) % 1000 == 0:\n",
    "                sys.stderr.write(\"Parsed %s introns...\\n\"%len(geneRows))\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    for i in fout:\n",
    "        fout[i].close()\n",
    "\n",
    "    # qqnorms on the columns\n",
    "    matrix = np.array(valRows)\n",
    "    for i in xrange(len(matrix[0,:])):\n",
    "        matrix[:,i] = qqnorm(matrix[:,i])\n",
    "        \n",
    "    # write the corrected tables\n",
    "    fout = {}\n",
    "    for i in chroms:\n",
    "        fn=\"%s.qqnorm_chr%s\"%(ratio_file,i)\n",
    "        print(\"Outputting: \" + fn)\n",
    "        fout[i] = file(fn,'w')\n",
    "        fout[i].write(\"\\t\".join(['#Chr','start','end','ID'] + header)+'\\n')\n",
    "    lst = []\n",
    "    for i in xrange(len(matrix)):\n",
    "        chrom, s = geneRows[i].split()[:2]\n",
    "        \n",
    "        lst.append((chrom.replace(\"chr\",\"\"), int(s), \"\\t\".join([geneRows[i]] + [str(x) for x in  matrix[i]])+'\\n'))\n",
    "\n",
    "    lst.sort()\n",
    "    for ln in lst:\n",
    "        fout[ln[0]].write(ln[2])\n",
    "        \n",
    "    fout_run = file(\"%s_prepare.sh\"%ratio_file,'w')\n",
    "\n",
    "    for i in fout:\n",
    "        fout[i].close()\n",
    "        fout_run.write(\"bgzip -f %s.qqnorm_chr%s\\n\"%(ratio_file, i))\n",
    "        fout_run.write(\"tabix -p bed %s.qqnorm_chr%s.gz\\n\"%(ratio_file, i))\n",
    "    fout_run.close()\n",
    "\n",
    "    sys.stdout.write(\"Use `sh %s_prepare.sh' to create index for fastQTL (requires tabix and bgzip).\\n\"%ratio_file)\n",
    "\n",
    "    if pcs>0:\n",
    "        #matrix = np.transpose(matrix) # important bug fix (removed as of Jan 1 2018)\n",
    "        pcs = min([len(header), pcs])\n",
    "        pca = PCA(n_components=pcs)                                                                                                                                                                            \n",
    "        pca.fit(matrix)  \n",
    "        pca_fn=ratio_file+\".PCs\"\n",
    "        print(\"Outputting PCs: \" + pca_fn)\n",
    "        pcafile = file(pca_fn,'w')  \n",
    "        pcafile.write(\"\\t\".join(['id']+header)+'\\n')\n",
    "        pcacomp = list(pca.components_)\n",
    "    \n",
    "        for i in range(len(pcacomp)):\n",
    "            pcafile.write(\"\\t\".join([str(i+1)]+[str(x) for x in pcacomp[i]])+'\\n')\n",
    "\n",
    "        pcafile.close()\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    parser = OptionParser(usage=\"usage: %prog [-p num_PCs] input_perind.counts.gz\")\n",
    "    parser.add_option(\"-p\", \"--pcs\", dest=\"npcs\", default = 50, help=\"number of PCs output\")\n",
    "    parser.add_option(\"--ChromosomeBlackList\", dest=\"cbl\", default=\"\", help=\"file of blacklisted chromosomes to exclude from analysis, one per line. If none is provided, will default to blacklisting X and Y\")\n",
    "    (options, args) = parser.parse_args()\n",
    "    if len(args)==0:\n",
    "        sys.stderr.write(\"Error: no ratio file provided... (e.g. python leafcutter/scripts/prepare_phenotype_table.py input_perind.counts.gz\\n\")\n",
    "        exit(0)\n",
    "    main(args[0], get_chromosomes(args[0]), get_blacklist_chromosomes(options.cbl), int(options.npcs) )\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
